### PR TITLE
Make error rendering context-aware

### PR DESCRIPTION
**Pull request by Gusted**
- Currently the error handling in `modules/context/context.go` always use `ctx.HTML`(indirectly) to render the error, however it is possible that the context doesn't have any HTML render set(e.g. `APIContext`).
- Let the context creator provide the error handling, currently web will still use the current functions and API/Package will roughly do the same but using `ctx.JSON`.
- Implement a `dummyRender` for the API/Package Context to not see a vague NPE error.

For clarity purposes, this panic was found via:

https://github.com/go-gitea/gitea/blob/6f3efdfe117b46978d449f9223394326e39067ce/routers/api/v1/api.go#L152 -> https://github.com/go-gitea/gitea/blob/6f3efdfe117b46978d449f9223394326e39067ce/modules/context/context.go#L144 -> https://github.com/go-gitea/gitea/blob/6f3efdfe117b46978d449f9223394326e39067ce/modules/context/context.go#L305 -> https://github.com/go-gitea/gitea/blob/6f3efdfe117b46978d449f9223394326e39067ce/modules/context/context.go#L227

ContextAPI doesn't set the `Render` field, which caused an NPE.

### CODE REVIEW

**Review by delvh — APPROVED** [review]

File: modules/context/api.go [hunk 1]
```diff
@@ -89,25 +93,19 @@ func (ctx *APIContext) ServerError(title string, err error) {
...
 		message = fmt.Sprintf("%s", obj)
 	}
 
 	if status == http.StatusInternalServerError {
-		log.ErrorWithSkip(1, "%s: %s", title, message)
-
-		if setting.IsProd && !(ctx.Doer != nil && ctx.Doer.IsAdmin) {
-			message = ""
-		}
+		log.ErrorWithSkip(1, "%s: %w", title, foundError)
```
Comment (zeripath) [inline modules/context/api.go]: %w doesn't work in log.

```suggestion
		log.ErrorWithSkip(1, "%s: %v", title, foundError)
```

**Comment by lunny** [comment]
For a long-term target, `APIContent` should have a new struct to remove unnecessary fields rather than deriving from `Context`. The benefit of that is less memory usage per request especially for those big Gitea instances. #16455 should be the first step towards that target. I'm not against this PR but I think this still need to be changed in future.

**Comment by 6543** [comment]
Without looking more into the code, did you make sure normal users dont see the stacktrace in prod mode?

**Comment by Gusted** [comment]
Stacktraces were never exposed to begin with and conditions were copies (ctrl + f `setting.IsProd`)

**Comment by lunny** [comment]
For web routers, sometimes it will return JSON, sometimes return HTML and for RSS it's XML.
I couldn't find it becomes more convenient than explicitly invoking a render function in the router function.

**Comment by Gusted** [comment]
> For web routers, sometimes it will return JSON, sometimes return HTML and for RSS it's XML. I couldn't find it becomes more convenient than explicitly invoking a render function in the router function.

That would require a very broad refactor, which I won't contribute. If that's the consensus to fix this problem, feel free to close my PR.

**Comment by wxiaoguang** [comment]
We need to decouple the Gitea's contexts first. Otherwise more things are mixed together.

->  Decouple WebContext/APIContext/PrivateContext #24786

**Pull request closed without merging**